validate emit results are defined in taskspec

Yongxuanzhang opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Task results are emitted in step script by writing to results path, before this commit we don't have validation of checking the results defined in taskspec. This commit adds the check in validation.

/kind bug

Signed-off-by: Yongxuan Zhang yongxuanzhang@google.com

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Tasks results emitted in script but don't defined in taskspec will fail
```

lbernick commented on pkg/apis/pipeline/v1/task_validation.go:
```diff
@@ -406,6 +407,17 @@ func validateTaskContextVariables(ctx context.Context, steps []Step) *apis.Field
 	return errs.Also(validateVariables(ctx, steps, "context\\.task", taskContextNames))
 }
 
+func validateTaskResultsVariables(ctx context.Context, steps []Step, results []TaskResult) (errs *apis.FieldError) {
```
Can you add a docstring for this function?

Yongxuanzhang replied:
sure! Thanks for the suggestion!
I found it hard to understand some functions in this file cuz they lack of docstrings

lbernick commented on pkg/apis/pipeline/v1/task_validation_test.go:
```diff
@@ -541,6 +568,26 @@ func TestTaskSpecValidateError(t *testing.T) {
 			Paths:   []string{"steps"},
 		},
+	}, {
+		name: "invalid results path variable in step script",
```
```suggestion
		name: "step script refers to nonexistent result",
```

Yongxuanzhang replied:
Thanks!

lbernick commented on pkg/apis/pipeline/v1/task_validation_test.go:
```diff
@@ -288,6 +288,33 @@ func TestTaskSpecValidate(t *testing.T) {
+			},
+		},
+	}, {
+		name: "valid results in step image",
```
should there also be a test case for an invalid result in an image?

pritidesai replied:
NIT: this test case is not necessary and might create confusion in future that a step `image` does accept `results`. This is anyways invalid, since `a-result` is not declared as a result.

Yongxuanzhang commented on pkg/apis/pipeline/v1/task_validation_test.go:
```diff
@@ -541,6 +568,26 @@ func TestTaskSpecValidateError(t *testing.T) {
+				#!/usr/bin/env bash
+				date | tee $(results.non-exist.path)`,
+			}},
+			Results: []v1.TaskResult{
+				{
+					Name: "a-result",
+				},
+			},
```
sure!

lbernick commented on pkg/apis/pipeline/v1/task_validation.go:
```diff
@@ -406,6 +407,18 @@ func validateTaskContextVariables(ctx context.Context, steps []Step) *apis.Field
 	return errs.Also(validateVariables(ctx, steps, "context\\.task", taskContextNames))
 }
 
+// validateTaskResultsVariables validates if the results reference in step script are defined in task results
```
```suggestion
// validateTaskResultsVariables validates if the results referenced in step script are defined in task results
```

Yongxuanzhang replied:
thanks!

pritidesai commented on pkg/apis/pipeline/v1beta1/task_validation_test.go:
```diff
@@ -324,6 +324,31 @@ func TestTaskSpecValidate(t *testing.T) {
+	}, {
+		// since $(results.resultName.path) cannot be used in fields other than script
+		// this is to make sure we don't check it.
+		name: "valid results  in step image",
```
same comment as ⬆️, this test case is not needed

pritidesai commented:
thank you @Yongxuanzhang, the PR looks great! Let me know if you have any strong opinion on keeping the test case, I am happy to merge either ways.

Yongxuanzhang commented:
> thank you @Yongxuanzhang, the PR looks great! Let me know if you have any strong opinion on keeping the test case, I am happy to merge either ways.

Thanks! I have removed this case.
Also added release note since this commit may be a change for some current users.

pritidesai commented:
thank you @Yongxuanzhang 👍 

/lgtm
